core/src/util/usLDAPExpr.cpp

```diff
@@ -0,0 +1,852 @@
+    for (std::size_t i = 0; i < d->m_args.size( ); i++)
+    {
+      LDAPExpr::ObjectClassSet r;
+      if (d->m_args[i].GetMatchedObjectClasses(r))
+      {
+        result = true;
+        if (objClasses.empty())
+        {
+          objClasses = r;
+        }
+        else
+        {
+          // if AND op and classes in several operands,
+          // then only the intersection is possible.
+          LDAPExpr::ObjectClassSet::iterator it1 = objClasses.begin();
+          LDAPExpr::ObjectClassSet::iterator it2 = r.begin();
+          while ( (it1 != objClasses.end()) && (it2 != r.end()) )
+          {
+            if (*it1 < *it2)
+            {
+              objClasses.erase(it1++);
+            }
+            else if (*it2 < *it1)
+            {
+              ++it2;
+            }
+            else
+            { // *it1 == *it2
+              ++it1;
+              ++it2;
+            }
+          }
+          // Anything left in set_1 from here on did not appear in set_2,
+          // so we remove it.
+          objClasses.erase(it1, objClasses.end());
+        }
+      }
+    }
+    return result;
+  }
+  else if (d->m_operator == OR)
+  {
+    for (std::size_t i = 0; i < d->m_args.size( ); i++)
+    {
+      LDAPExpr::ObjectClassSet r;
+      if (d->m_args[i].GetMatchedObjectClasses(r))
+      {
+        std::copy(r.begin(), r.end(), std::inserter(objClasses, objClasses.begin()));
+      }
+      else
+      {
+        objClasses.clear();
+        return false;
+      }
+    }
+    return true;
+  }
+  return false;
+}
+
+std::string LDAPExpr::ToLower(const std::string& str)
+{
+  std::string lowerStr(str);
+  std::transform(str.begin(), str.end(), lowerStr.begin(), ::tolower);
+  return lowerStr;
+}
+
+bool LDAPExpr::IsSimple(const StringList& keywords, LocalCache& cache,
+                        bool matchCase ) const
+{
+  if (cache.empty())
+  {
+    cache.resize(keywords.size());
+  }
+
+  if (d->m_operator == EQ)
+  {
+    StringList::const_iterator index;
+    if ((index = std::find(keywords.begin(), keywords.end(), matchCase ? d->m_attrName : ToLower(d->m_attrName))) != keywords.end() &&
+        d->m_attrValue.find_first_of(LDAPExprConstants::WILDCARD()) == std::string::npos)
+    {
+      cache[index - keywords.begin()] = StringList(1, d->m_attrValue);
+      return true;
+    }
+  }
+  else if (d->m_operator == OR)
+  {
+    for (std::size_t i = 0; i < d->m_args.size( ); i++)
+    {
+      if (!d->m_args[i].IsSimple(keywords, cache, matchCase))
+        return false;
+    }
+    return true;
+  }
+  return false;
+}
+
+bool LDAPExpr::IsNull() const
+{
+  return !d;
+}
+
+bool LDAPExpr::Query( const std::string& filter, const Properties& pd)
+{
+  return LDAPExpr(filter).Evaluate(pd, false);
+}
+
+bool LDAPExpr::Evaluate( const Properties& p, bool matchCase ) const
+{
+  if ((d->m_operator & SIMPLE) != 0)
+  {
+    // try case sensitive match first
+    int index = p.FindCaseSensitive(d->m_attrName);
```

Comment:
This was optimized for the general case were it is assumed that people use filters on entities that contain the attribute names as property keys with the same case and the filter compares property values most of the time. Case mismatches are assumed to be typos or "laziness".